pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,178 @@
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// List all the Namespaces which matches the Namespace LabelSelector.
+	namespaceList := corev1.NamespaceList{}
+	if ingressController.Spec.NamespaceSelector != nil {
+		namespaceLabelSelector := ingressController.Spec.NamespaceSelector.MatchLabels
+		if err := r.cache.List(ctx, &namespaceList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(namespaceLabelSelector),
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Namespaces %q: %v", request, err)
+		}
+	} else {
+		// If Namespace LabelSelector is not provided then add Namespace with empty name to indicate all Namespaces.
+		namespaceList.Items = append(namespaceList.Items, corev1.Namespace{})
+	}
+
+	// Get the Route LabelSelector from the Ingress Controller.
+	var routeLabelSelector map[string]string
+	if ingressController.Spec.RouteSelector != nil {
+		routeLabelSelector = ingressController.Spec.RouteSelector.MatchLabels
+	}
+
+	// Variable to store the number of routes admitted by the Shard (Ingress Controller).
+	routesAdmitted := 0
+
+	// Iterate through all the Namespaces.
+	for _, namespace := range namespaceList.Items {
+		// List all the Routes which matches the Route LabelSelector and the Namespace.
+		routeList := routev1.RouteList{}
+		if err := r.cache.List(ctx, &routeList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(routeLabelSelector),
+			Namespace:     namespace.Name,
+		}); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list Routes for the Shard %q: %v", request, err)
+		}
+
+		// Iterate through all the Routes.
+		for _, route := range routeList.Items {
+			// Check if the Route was admitted by the Ingress Controller.
+			if routeAdmitted(route, ingressController.Name) {
+				// If the Route was admitted then, the routesAdmitted should be incremented by 1 for the Shard.
+				routesAdmitted++
+			}
+		}
+	}
+
+	// Set the value of the metric to the number of routesAdmitted for the corresponding Shard (Ingress Controller).
+	SetRouteMetricsControllerRoutesPerShardMetric(request.Name, float64(routesAdmitted))
+
+	return reconcile.Result{}, nil
+}
+
+// routeAdmitted returns true if a given route has been admitted by the Ingress Controller.
```

Comment:
It doesn't matter which controller "admitted" the route, so in this code you can just check the route status.
```suggestion
// routeAdmitted returns true if a given route has been admitted 
```

Reply:
We are checking if a Route is admitted by a particular Ingress Controller as only then we would be able to get the correct value for `RoutesPerShard` for this Ingress Controller. If we don't check which Ingress Controller admitted the Route then we will end up with incorrect value for the metric.